fixes #3492 - API v2 nested routes for each controller

isratrade opened the pull request.

isratrade commented on app/views/api/v2/operatingsystems/show.json.rabl:
```diff
@@ -1,22 +1,24 @@
 
 child :media, :object_root => false do
-  attributes :id, :name
+  extends "api/v2/media/base"
```
@domcleal, he is an example of extending a `base` template on a child node.

domcleal replied:
ah, thanks.  I like the de-duplication.

pitr-ch commented on config/routes/api/v2.rb:
```diff
@@ -78,6 +85,15 @@
+        resources :puppetclasses, :except => [:new, :edit]
+        resources :config_templates, :except => [:new, :edit]
+        resources :os_default_templates, :except => [:new, :edit]
+        resources :trends, :except => [:new, :edit]
```
I think there is a helper for this, `api_resource` or something like that, not sure if it is in katello in foreman. @tstrachota?

tstrachota replied:
That was in katello https://github.com/Katello/katello/blob/9448155494d1c33a99ab2a38c151c6de6b31afaf/app/lib/api/mapper_extensions.rb#L6
I consider it safer to use a helper.

domcleal commented on app/controllers/api/v2/hostgroups_controller.rb:
```diff
@@ -5,17 +5,25 @@ class HostgroupsController < V2::BaseController
       api :GET, "/hostgroups/", "List all hostgroups."
+      api :GET, "/operatingsystems/:operatingsystem_id/hostgroups", "List all hostgroups for an operating system"
+      api :GET, "/config_templates/:config_template_id/hostgroups", "List all hostgroups for a provisioning template"
+      api :GET, "/domains/:domain_id/hostgroups", "List all hostgroups for a domain"
```
missing config_template_id and domain_id from apidoc below

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -198,6 +205,12 @@ def detect_host_type
+      private
+
+      def allowed_nested_id
+        %w(operatingsystem_id compute_resource_id domain_id)
```
domain_id isn't in the apidoc above

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -7,17 +7,24 @@ class HostsController < V2::BaseController
       add_puppetmaster_filters :facts
 
       api :GET, "/hosts/", "List all hosts."
+      api :GET, "/operatingsystems/:operatingsystem_id/hosts", "List all hosts for an operating system"
```
Aren't there many other resources associated with hosts, like media, ptables etc?  Should we leave them as they're slightly strange combinations, or add as many as we can?

isratrade replied:
@domcleal, since nested routes and just another way to search, I agree that we don't need to include every association.  The question is what is 'strange' and what is not.

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -73,7 +73,7 @@ def destroy
 
       def allowed_nested_id
-        %w(host_id)
+        %w(host_id domain_id)
```
not in the apidoc above?  subnet_id too if you're doing this?

domcleal commented on app/controllers/api/v2/subnets_controller.rb:
```diff
@@ -5,17 +5,24 @@ class SubnetsController < V2::BaseController
       param :order, String, :desc => 'Sort results'
       param :page, String, :desc => "paginate results"
       param :per_page, String, :desc => "number of entries per request"
+      param :domain_id, String, :desc => "id of nested subnet"
```
s/subnet/domain/

domcleal commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -96,10 +105,17 @@ def destroy
+      private
+
+      def allowed_nested_id
+        %w(auth_source_ldap_id compute_resource_id domain_id)
```
remove domain_id?

domcleal commented:
Can we de-duplicate this a bit, e.g.

<pre>
        @architectures = nested_obj ? nested_obj.architectures : resource_scope
        @architectures = @architectures.search_for(*search_options).paginate(paginate_options)
</pre>

A helper for the routes sounds good too.

I'm curious what you want to do with hosts and host groups, since they link to many objects, and whether we want to add routes for those too.

pitr-ch commented:
Turning on api-pie param validation could be helpful too. It would help to reveal many of these documentation-vs-reality inconsistencies.

isratrade commented:
@domcleal, thanks for your feedback.  This PR is still a work in process.  I moved DO NOT MERGE, from the description to the title.

domcleal commented:
@isratrade don't worry, I'd noticed :)  Was just trying to give you some early feedback before I'm on PTO.

dLobatog commented:
@isratrade ping. Can you rebase this?

isratrade commented:
@eLobato, can you please review this?

domcleal commented:
I'd like to give this another review too, so please don't merge immediately.

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -17,10 +20,19 @@ class ConfigTemplatesController < V2::BaseController
-          authorized(:view_templates).
-          search_for(*search_options).paginate(paginate_options).
-          includes(:operatingsystems, :template_combinations, :template_kind)
+        if nested_obj
```
I think we can generalize this to:

``` ruby
all_config_templates = (nested_obj.present? ? nested_obj.config_templates : ConfigTemplate)
@config_templates = all_config_templates.
                                          authorized(:view_templates).
                                          includes(:operatingsystems, :template_combinations, :template_kind).
                                          search_for(*search_options).paginate(paginate_options)
@total = all_config_templates.authorized(:view_architectures).count
```

and avoid the if clause.

dLobatog replied:
I just finished reading the rest, I think the above code should be refactored in the api/base_controller and apply it to all controllers that require it via a before_filter. At the end of the day it's very similar code duplicated in all controllers and it'd be easier to maintain and cleaner if it's all kept in the base_controller for v2.

isratrade replied:
@eLobato, yes it's similar code that can be DRY'd up, but the problem is that `all_config_templates` will return the entire array, not paginated, which is not good.  The pagination and search should be all the same db retrieval call.

dLobatog replied:
Hmm what about something along the lines of:

``` ruby
all_config_templates = (nested_obj.present? ? "nested_obj.config_templates" : ConfigTemplate)
@config_templates = eval(all_config_templates).
                                          authorized(:view_templates).
                                          includes(:operatingsystems, :template_combinations, :template_kind).
                                          search_for(*search_options).paginate(paginate_options)
```

I hope there is some elegant way of delaying the db call (I'd say 'eval' in production code looks like a hack), but in the worst case this would put both pagination and search in the same db retrieval call.

isratrade replied:
@eLobato, `nested_obj.send(:config_template)` is safe than eval,

dLobatog replied:
True, but wouldn't `.send(:config_templates)` make the db call right away?

pitr-ch replied:
depends on what is in `nested_obj`, if `config_templates` is just association it'll be lazy.

pitr-ch replied:
This could be DRY-ed:

``` ruby
config_templates_scope = nested_obj ? nested_obj.config_templates : ConfigTemplate.scoped
@config_templates      = config_templates_scope.
    authorized(:view_templates).
    includes(:operatingsystems, :template_combinations, :template_kind).
    search_for(*search_options).
    paginate(paginate_options)
@total                 = config_templates_scope.authorized(:view_templates).count
```

pitr-ch replied:
I just notices that is already mentioned in https://github.com/theforeman/foreman/pull/1068#discussion_r12085491. The issue discussed here regarding all_config_templates containing all the config templates in Array is based on false assumption. `all_config_templates` or in my example called `config_templates_scope` will contain scope not Array and will not trigger SQL queries. I am assuming `nested_obj.config_templates` is calling regular association.

isratrade replied:
@pitr-ch, good to know. I thought the array was returned but you're right, the `scoped` and relationship were not trigger SQL queries right away. I'll DRY things up.

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -17,10 +20,19 @@ class ConfigTemplatesController < V2::BaseController
+                             authorized(:view_templates).
+                             includes(:operatingsystems, :template_combinations, :template_kind).
+                             search_for(*search_options).paginate(paginate_options)
+          @total = nested_obj.config_templates.authorized(:view_architectures).count
```
I think this `:view_architectures` should be `view_templates`

isratrade replied:
@eLobato, fixed. thanks for catching.

isratrade commented:
I prefer the non-DRY, but more efficient, code that I have in this PR. Any blockers to merging this PR?

isratrade commented:
stage is "needs re-review".  Any takers?

isratrade commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -7,19 +7,30 @@ class HostsController < V2::BaseController
 
       def index
-        @hosts = resource_scope.search_for(*search_options).paginate(paginate_options)
+        hosts_scope = nested_obj ? nested_obj.hosts.authorized(:view_hosts) : resource_scope
```
I used `resource_scope` and moved `authorized(:view_hosts)` to the link of `hosts_scope =` since there was a test failure if it was defined like the others.

isratrade commented:
@domcleal, @pitr-ch, can you please review my DRY changes

pitr-ch commented:
I would go further, the index methods are essentially all the same. It would be better to share that through inheritance.

isratrade commented:
I don't recall other places in the code where we abstract controller methods. You're basically saying to expand `resource_scope` in api/base_controller.rb.  I'm not sure if it saves much.  Some controllers have `includes()`, other's don't, so these needs to be passed as a variable. Most times, the controller name is the permission, but sometimes it's not (like view_templates), so there's to be an override. In short, I'm not sure it helps with the readable of code.

pitr-ch commented:
I see it from the other side the pattern is repeated about 13times and it differs only in resource class (association name can be inferred most of the time) includes and name of the permission. 13times makes it very probable that next time somebody will only change one of these places and inconsistency and bugs are born. I think following would be readable:

``` ruby
class AnAbstractController
  def get_resources_for_index(resource_class, options)
    raise KeyError unless (options.keys - [:association, :permission, :includes]).empty?

    association_name = options[:association] || recource_class.name.underscore.plularize
    # TODO interpolate permission name if not supplied
    resource_scope   = nested_obj ? nested_obj.send(association_name) : resource_class.scoped
    resources        = resource_scope.
        authorized(options[:permission]).# returns self when :permission nil
        includes(*Array(options[:includes])).# allows nil, [], one includes or more in an array
        search_for(*search_options).
        paginate(paginate_options)
    [resources, resource_scope.count]
  end
end

class SubnetController < AnAbstractController
  def index
    @subnets, @total = get_resources_for_index Subnet,
                                               permission: :view_subnets,
                                               includes:   [:tftp, :dhcp, :dns]
  end
end

class ConfigTemplateController < AnAbstractController
  def index
    @config_templates, @total = get_resources_for_index(
        ConfigTemplate,
        permission: :view_templates,
        includes:   [:operatingsystems, :template_combinations, :template_kind])
  end
end

# etc.
```

dLobatog commented:
Quick note, `V2::BaseController` could very well do the job of `AbstractController` in the example above

pitr-ch commented:
@eLobato yeah it could, the names was just to demonstrate ...

isratrade commented:
@pitr-ch, in your example, it is not correct to assign `@subnets, @total =`.
`@total =` doesn't use scoped search in for `.count` and doesn't need the `includes()` queries.

isratrade commented:
@pitr-ch, ignore above. I now understand what you did with the multiple assignment using `@total`.  The only issue I see is that it should add the permission such as `resource_scope.authorized(options[:permission]).count`

pitr-ch commented:
Yeah, I've missed that. I was putting it together in a hurry.

isratrade commented:
@pitr-ch, I pushed changes to `api/base_controller` and `domains_controller`. I wanted to get your feedback, before I update the other controllers.
In short, I decided to use the expand the existing `resource_scope` method in `base_controller` which can be used whether or not there is a nested route. I did not DRY `.search` and `.paginate` for each controller as it's not consistent with the UI controllers. Also, now that `.search` is not part of `resource_scope`, I don't need to explicitly assign `@total =` since we can use this
https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/base_controller.rb#L27-29

pitr-ch commented:
Nice! :) just two minor comments.

isratrade commented:
ok, will add `def resource_scope_for_index`

isratrade commented:
@pitr-ch, can you review? I squashed all my changes.

isratrade commented on app/controllers/api/base_controller.rb:
```diff
@@ -48,17 +48,48 @@ def resource_class
+      if nested_obj
+        if nested_obj.respond_to?(options[:association]) && nested_obj.send(options[:association]).respond_to?(:authorized)
+          nested_obj.send(options[:association]).authorized(options[:permission], resource_class)
+        elsif nested_obj.respond_to?(options[:association])
```
I added this check here, since `interfaces` don't `respond_to?(:authorized)`

isratrade commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,22 +5,28 @@ class ConfigTemplatesController < V2::BaseController
-          authorized(:view_templates).
-          search_for(*search_options).paginate(paginate_options).
-          includes(:operatingsystems, :template_combinations, :template_kind)
+        @config_templates = resource_scope_for_index('config_templates', :permission => :view_templates).includes(:template_kind)
```
I don't see the reason to include `operatingsystems, :template_combinations` in the `index` since this data is not in the response.

isratrade commented on app/controllers/api/v2/hostgroups_controller.rb:
```diff
@@ -5,19 +5,23 @@ class HostgroupsController < V2::BaseController
       def index
-        @hostgroups = Hostgroup.
-          authorized(:view_hostgroups).
-          includes(:hostgroup_classes, :group_parameters).
```
I didn't include `hostgroup_classes, :group_parameters` since this data is not in the response

isratrade commented on app/controllers/api/v2/operatingsystems_controller.rb:
```diff
@@ -6,19 +6,25 @@ class OperatingsystemsController < V2::BaseController
-          authorized(:view_operatingsystems).
-          includes(:media, :architectures, :ptables, :config_templates, :os_default_templates).
-          search_for(*search_options).paginate(paginate_options)
+        @operatingsystems = resource_scope_for_index
```
same, no `includes(:media, :architectures, :ptables, :config_templates, :os_default_templates)` since data not used in response

pitr-ch replied:
:+1: thanks for checking and removing that

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -48,17 +48,48 @@ def resource_class
+
+    def resource_scope_for_index(controller = controller_name, options = {})
+      resource_scope(controller, options).search_for(*search_options).paginate(paginate_options)
+    end
```
By including the controller var into options, the names of controllers would not have to be repeated when options needs to be used.

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -48,17 +48,48 @@ def resource_class
+    end
+
+    def resource_scope_nested(controller = controller_name, options = {})
+      raise KeyError unless (options.keys - [:association, :permission, :includes]).empty?
```
this check also applies in resource_scope_not_nested, it can be moved to #resource_scope

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -48,17 +48,48 @@ def resource_class
+    def resource_scope_not_nested(controller = controller_name, options = {})
+      scope = resource_class.scoped
+      options[:permission] ||= "#{action_permission}_#{controller}"
+      if resource_class.respond_to?(:authorized)
```
this could lead to silently ignored :permission option, it would be better to raise when :permission is supplied but not supported by the resource_class

isratrade replied:
@pitr-ch, @ares, is there a helper method to check when :permission is supplied but not supported by the resource_class? Since this is code and not use input, I don't see a big value in this check. Our code is not going to have `resource_scope_for_index(:permission => 'foobar')`

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -48,17 +48,48 @@ def resource_class
+        elsif nested_obj.respond_to?(options[:association])
+          nested_obj.send(options[:association])
+        else
+          resource_scope_not_nested(controller, options)
```
same here, this branch of `ifs` can silently ignore :association option, it would be better to restructure it here in a way that options is authoritative in what methods resource has to support.

isratrade replied:
@pitr-ch, can you explain further here

isratrade commented:
@pitr-ch, I re-factored so `controller` as an arguement is remove, and it is not part of `options`

isratrade commented on app/controllers/api/v2/common_parameters_controller.rb:
```diff
@@ -11,10 +11,7 @@ class CommonParametersController < V2::BaseController
-          authorized(:view_globals, CommonParameter).
-          search_for(*search_options).
-          paginate(paginate_options)
+        @common_parameters = resource_scope_for_index(:permission => :view_globals)
```
@domcleal, verify this change doesn't affect the changes you made in 7cb05aa94e942bd3917c6cde33957288ea84a735

domcleal replied:
untested, but looks like it will.  You might need to override `resource_class` in this controller to return CommonParameter.

isratrade replied:
@domcleal, it should be OK, since the controller name is `common_parameters` so the `resource_class` will be CommonParameter

domcleal replied:
Ah yes, you're right.

The issue I'd had was because the class was derived from within the authorized scope, and since the scope was included in Parameter, it returned Parameter rather than the CommonParameter subclass.  Gah.

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -45,20 +45,54 @@ def resource_name
-    def resource_scope(controller = controller_name)
+    def resource_scope(options = {})
+      if (k = (options.keys - [:controller, :association, :permission, :includes])).any?
+        raise ::Foreman::Exception.new(N_("unknown option for %s"), k.join(', ')) and return
```
return will be never called

pitr-ch commented on app/controllers/api/base_controller.rb:
```diff
@@ -45,20 +45,54 @@ def resource_name
+      else
+        resource_scope_not_nested(options)
       end
     end
```
This would be more readable

``` ruby
def resource_scope_nested(options)
  if nested_obj && nested_obj.respond_to?(options[:association])
    association = nested_obj.send(options[:association])
    if association.respond_to?(:authorized)
      association.authorized(options[:permission], resource_class)
    else
      association
    end
  else
    resource_scope_not_nested(options)
  end
end
```

dLobatog replied:
Fixed, Github didn't wrap your comment for some reason

pitr-ch commented:
Two minor comments otherwise ack. Thanks.

isratrade commented:
[test]

isratrade commented:
@pitr-ch, re-based with your suggested fixes.

dLobatog commented on app/controllers/concerns/api/v2/taxonomies_controller.rb:
```diff
@@ -2,11 +2,11 @@ module Api::V2::TaxonomiesController
   extend ActiveSupport::Concern
 
   included do
+    before_filter :find_optional_nested_object
```
I think this misses `:only => %w(index)`

isratrade replied:
@eLobato, if you notice, all the controllers with rested routes have `before_filter :find_optional_nested_object` that are on all actions, not just `:index`
This is to catch any potential errors in case the the nested :id doesn't exisit.

pitr-ch commented:
ACK

isratrade commented:
refer to PR #1454 for nested routes to add/remove many-to-many associations

isratrade commented:
[test]

isratrade commented:
@domcleal, can this PR be merged in 1.5.x? Adding nested routes to API version that is still experimental shouldn't have a major impact.

domcleal commented:
I'm unsure about merging into 1.5-stable, there are quite a lot of changes to the base controller and how resources are fetched that probably need some time testing in develop before moving to a stable branch in case bugs exist.  I'd say probably not for .1, perhaps .2 if you think it's important we do so.

isratrade commented:
@domcleal, you're right. There are a lot of changes to api/base_controller. If it only affected v2 (non-default), I would say to merge it, but it can wait until 1.6-develop

isratrade commented on app/controllers/api/v2/architectures_controller.rb:
```diff
@@ -1,19 +1,19 @@
-          authorized(:view_architectures).
-          includes(:operatingsystems).
-          search_for(*search_options).paginate(paginate_options)
+        @architectures = resource_scope_for_index
```
@domcleal, @ohadlevy, Is this PR with so many changes, such as the one above, re-factoring to use `resource_scope_for_index` be something that can be merged after v2 was declared 'stable'?

domcleal replied:
Is there any effect on the API user?

isratrade replied:
@domcleal, there should not be any effect ton the API user, so only 2nd thought, I think it will be good to merge this PR as it adds routes and cleans/DRYs up the code.  @eLobato, @pitr-ch, do you mind reviewing again?

dLobatog commented on app/controllers/api/base_controller.rb:
```diff
@@ -43,20 +43,53 @@ def resource_name
+        raise ::Foreman::Exception.new(N_("unknown option for %s"), k.join(', '))
+      end
+
+      options[:association] ||= controller_name
```
Why two variables with the same value?

dLobatog commented:
Tests are failing, I'll test and re-review after those are fixed

isratrade commented:
@eLobato, fixed failing api/users tests. will squash commit after review and tests pass.

isratrade commented on app/controllers/concerns/foreman/controller/users_mixin.rb:
```diff
@@ -7,7 +7,7 @@ module Foreman::Controller::UsersMixin
 
   def resource_scope(controller = controller_name)
-    super(controller).except_hidden
+    super(:controller => controller).except_hidden
```
@eLobato, all tests pass after making this change here on user_mixin.rb

dLobatog commented:
[test] , the job got aborted during the Jenkins deadlock

dLobatog commented:
Tested it and it works nicely. Maybe we need support for nested routes in external user groups/auth sources and external user groups/user groups. @domcleal wanted to gave this the last review if I remember correctly, please go ahead.

isratrade commented:
@eLobato, what would the new v2 routes be for external groups? There is no `/external_usergroups` in the UI. In the API, there is `GET usergroups/:id/external_usergroups`. Would you add 3rd level deep route for
 `GET usergroups/:usergroup_id/external_usergroups/:id/auth_sources` ?

isratrade commented:
@eLobato, re-based and all tests passing. Do you want to add /external_usergroups routes as separate PR or part of this PR?

pitr-ch commented:
@isratrade looks good to me, but I am out of the loop so consider this half-ACK :)

isratrade commented:
@eLobato, ping. I don't understand the nested routes need for external usergroups. Since there is one AuthSource and one Usergroup per ExternalUsergroup, there is no collection to show in a nested route. `api/external_usergroups/:id doesn't yet exist, but if it did, it the`show` template should show it's related usergroup and auth_source.

dLobatog commented:
@isratrade the routes I meant would be the opposite:

`auth_sources/:id/external_usergroups` - shows all external user groups managed by an auth source
`usergroups/:id/external_usergroups` - shows all external user groups belonging to an user group.

dLobatog commented:
Since we have now `hostgroups/:hostgroup_id/hosts`, would it be much of an effort to add `environments/:environment_id/hosts`?

Commit:
fixes #7584 - add Authorizable module to class AuthSource

isratrade commented:
@eLobato, 
1) updated PR per your suggestions in a separate commit (to squash) to make it easier for review.
2) added `include Authorizable` to AuthSource as `find_nested_object` failed for the nested route `auth_source_ldaps/:id/external_usergroups` without it.
3) added route `/environments/:environment_id/hosts`

dLobatog commented:
Ready for merge after fixing the bug on hosts_controller and when tests are green :+1:

isratrade commented:
[test]

dLobatog commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -5,22 +5,30 @@ class HostsController < V2::BaseController
+      param :hostgroup_id, String, :desc => N_("id of nested hostgroup")
+      param :location_id, String, :desc => N_("id of nested location")
+      param :organization_id, String, :desc => N_("id of nested organization")
+      param :organization_id, String, :desc => N_("id of nested environment")
```
This should be `param :environment_id` , I think that's why tests are failing.

isratrade commented:
@eLobato, rebased

isratrade commented:
[test]

isratrade commented:
all tests passing. only `rundeck` failure on 1.8.7 that can be ignored and will be fixed by PR https://github.com/theforeman/foreman/pull/1771

dLobatog commented:
[test]

dLobatog commented:
@isratrade Looks ready to merge for me :+1: just get rid of the options[:association] or options[:controller] variables duplicity.

isratrade commented:
@eLobato, I moved options[:associationc] in this commit to be squashed. 017a9eef0364809183e2f93114e7c7aa243654c
I remember there was a reason that `:association` as an options key even though I see that I didn't use it.

dLobatog commented:
@isratrade Great, thanks. It needs a rebase now though, looks ready to merge for me after that. Why is #7584 in though?

domcleal commented:
API description strings look very inconsistent, please review them and ensure they use the English descriptions of resources to match the rest of the docs, not the model names.  e.g. "operating system", "provisioning template", "ID" should be capitalised.

isratrade commented:
PR #7584 came about because of this PR of nested routes. There were failing tests because of no authorization on AuthSource

isratrade commented:
on a high level, any ideas why this now have many test failures after rebase?

isratrade commented:
@eLobato, quite a lot of re-factoring and fixes after rebasing after the `FindCommon` mixin was merged. Changes are in separate commit (to later squash) to help review.

isratrade commented on app/controllers/api/v1/hosts_controller.rb:
```diff
@@ -125,8 +125,8 @@ def status
-      def resource_scope(controller = controller_name)
-        Host.authorized("#{action_permission}_#{controller}", Host)
+      def resource_class
+        Host::Managed
```
Only need to overwrite `resource_class` here, not `resource_scope`

```
> Host.respond_to?(:authorized)
 => false 
> Host::Managed.respond_to?(:authorized)
 => true 
```

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,20 +5,23 @@ class ConfigTemplatesController < V2::BaseController
       before_filter :process_operatingsystems, :only => [:create, :update]
 
       api :GET, "/config_templates/", N_("List provisioning templates")
+      api :GET, "/operatingsystem/:operatingsystem_id/config_templates", N_("List templates per operatingsystem")
```
`operatingsystem` --> `operating system`

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,20 +5,23 @@ class ConfigTemplatesController < V2::BaseController
+      api :GET, "/operatingsystem/:operatingsystem_id/config_templates", N_("List templates per operatingsystem")
+      api :GET, "/locations/:location_id/config_templates/", N_("List templates per location")
+      api :GET, "/organizations/:organization_id/config_templates/", N_("List templates per organization")
+      param :operatingsystem_id, String, :desc => N_("id of nested operatingsystem")
```
`ID` should be capitalized, also `operating system`

dLobatog commented on app/controllers/api/v2/domains_controller.rb:
```diff
@@ -16,16 +16,19 @@ class DomainsController < V2::BaseController
+      api :GET, "/subnets/:subnet_id/domains", N_("List of domains per subnet")
+      api :GET, "/locations/:location_id/domains", N_("List of domains per location")
+      api :GET, "/organizations/:organization_id/domains", N_("List of domains per organization")
+      param :subnet_id, String, :desc => N_("id of nested subnet")
```
`id` ---> `ID`

dLobatog commented on app/controllers/api/v2/environments_controller.rb:
```diff
@@ -5,16 +5,20 @@ class EnvironmentsController < V2::BaseController
+      api :GET, "/puppetclasses/:puppetclass_id/environments", N_("List environments of puppetclass")
+      api :GET, "/locations/:location_id/environments", N_("List environments per location")
+      api :GET, "/organizations/:organization_id/environments", N_("List environments per organization")
+      param :puppetclass_id, String, :desc => N_("id of nested puppetclass")
```
`id` ---> `ID`

dLobatog replied:
`puppetclass` ---> `Puppet class`

isratrade replied:
@eLobato, why to capitalize "Puppet class" when other words are not capitalized? i.e. environment, domain, etc

dLobatog commented on app/controllers/api/v2/environments_controller.rb:
```diff
@@ -5,16 +5,20 @@ class EnvironmentsController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy}
 
       api :GET, "/environments/", N_("List all environments")
+      api :GET, "/puppetclasses/:puppetclass_id/environments", N_("List environments of puppetclass")
```
`puppetclass` ---> `Puppet class`

dLobatog commented on app/controllers/api/v2/hostgroups_controller.rb:
```diff
@@ -5,17 +5,19 @@ class HostgroupsController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy clone}
 
       api :GET, "/hostgroups/", N_("List all host groups")
+      api :GET, "/puppetclasses/:puppetclass_id/hostgroups", N_("List all hostgroups for a puppetclass")
```
`host group` `Puppet class` `ID`, same in the other changes in this file.

dLobatog commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -5,19 +5,27 @@ class HostsController < V2::BaseController
       add_puppetmaster_filters :facts
 
       api :GET, "/hosts/", N_("List all hosts")
+      api :GET, "/hostgroups/:hostgroup_id/hosts", N_("List all hosts for a hostgroup")
```
`host group`, capitalized`ID`, same in params and other apipie docs in this file.

dLobatog commented on app/controllers/api/v2/images_controller.rb:
```diff
@@ -1,22 +1,29 @@
       api :GET, "/compute_resources/:compute_resource_id/images/", N_("List all images for a compute resource")
+      api :GET, "/operatingsystems/:operatingsystem_id/images/", N_("List all images for operating system")
+      api :GET, "/architectures/:architecture_id/images/", N_("List all images for architecture")
+      param :compute_resource_id, :identifier, :required => false
```
Why are these parameters not required?

dLobatog commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -5,15 +5,20 @@ class InterfacesController < V2::BaseController
+      api :GET, '/domains/:domain_id/interfaces', N_('List all interfaces for domain')
+      api :GET, '/subnets/:subnet_id/interfaces', N_('List all interfaces for subnet')
       param :host_id, String, :required => true, :desc => N_("ID or name of host")
+      param :domain_id, String, :required => false, :desc => N_('id or name of domain')
```
``` ID```
```

dLobatog commented on app/controllers/api/v2/media_controller.rb:
```diff
@@ -21,13 +22,15 @@ class MediaController < V2::BaseController
+      api :GET, "/operatingsystems/:operatingsystem_id/media", N_("List all media for an operating system")
+      api :GET, "/locations/:location_id/media", N_("List all media per location")
+      api :GET, "/organizations/:organization_id/media", N_("List all media per organization")
+      param :operatingsystem_id, String, :desc => N_("id of nested operatingsystem")
```
`ID of nested operating system`

dLobatog commented on app/controllers/api/v2/operatingsystems_controller.rb:
```diff
@@ -6,16 +6,22 @@ class OperatingsystemsController < V2::BaseController
+      api :GET, "/architectures/:architecture_id/operatingsystems", N_("List all operating systems for nested architecture")
+      api :GET, "/media/:medium_id/operatingsystems", N_("List all operating systems for nested medium")
+      api :GET, "/ptables/:ptable_id/operatingsystems", N_("List all operating systems for nested ptable")
+      api :GET, "/config_templates/:config_template_id/operatingsystems", N_("List all operating systems for nested config_template")
```
`config_template` -> `template` or `config template`, same in the params below.

dLobatog commented on app/controllers/api/v2/operatingsystems_controller.rb:
```diff
@@ -6,16 +6,22 @@ class OperatingsystemsController < V2::BaseController
+      api :GET, "/media/:medium_id/operatingsystems", N_("List all operating systems for nested medium")
+      api :GET, "/ptables/:ptable_id/operatingsystems", N_("List all operating systems for nested ptable")
+      api :GET, "/config_templates/:config_template_id/operatingsystems", N_("List all operating systems for nested config_template")
+      param :architecture_id, String, :desc => N_("id of nested architecture")
```
Capitalize `ID` in all param docs.

dLobatog commented on app/controllers/api/v2/os_default_templates_controller.rb:
```diff
@@ -9,12 +9,13 @@ class OsDefaultTemplatesController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy}
 
       api :GET, '/operatingsystems/:operatingsystem_id/os_default_templates', N_('List default templates combinations for an operating system')
+      api :GET, '/config_templates/:config_template_id/os_default_templates', N_('List os default templates for config template')
```
`os` --> `OS``

dLobatog commented on app/controllers/api/v2/ptables_controller.rb:
```diff
@@ -1,15 +1,17 @@
       before_filter :find_resource, :only => %w{show update destroy}
 
       api :GET, "/ptables/", N_("List all partition tables")
+      api :GET, "/operatingsystems/:operatingsystem_id/ptables", N_("List all ptables for an operating system")
```
`ptables` -> `partition tables`

dLobatog commented on app/controllers/api/v2/ptables_controller.rb:
```diff
@@ -1,15 +1,17 @@
 
       api :GET, "/ptables/", N_("List all partition tables")
+      api :GET, "/operatingsystems/:operatingsystem_id/ptables", N_("List all ptables for an operating system")
+      param :operatingsystem_id, String, :desc => N_("id of nested operatingsystem")
```
`ID of operating system`

dLobatog commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -8,14 +8,23 @@ class UsersController < V2::BaseController
+      before_filter :find_resource, :only => [:show, :update, :destroy]
 
       api :GET, "/users/", N_("List all users")
+      api :GET, "/auth_source_ldaps/:auth_source_ldap_id/users", N_("List all users for auth source ldap")
```
Capitalize `LDAP`

dLobatog commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -8,14 +8,23 @@ class UsersController < V2::BaseController
 
       api :GET, "/users/", N_("List all users")
+      api :GET, "/auth_source_ldaps/:auth_source_ldap_id/users", N_("List all users for auth source ldap")
+      api :GET, "/usergroups/:usergroup_id/users", N_("List all users for usergroup")
```
`user group` instead of `usergroup`

dLobatog commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -8,14 +8,23 @@ class UsersController < V2::BaseController
+      api :GET, "/roles/:role_id/users", N_("List all users for role")
+      api :GET, "/locations/:location_id/users", N_("List all users for location")
+      api :GET, "/organizations/:organization_id/users", N_("List all users for organization")
+      param :auth_source_ldap_id, String,  :desc => N_("id of nested auth source ldap")
```
Capitalize `ID` and `LDAP`, no need to mention 'nested' in api docs.

dLobatog commented:
Thanks @isratrade, API looks cleaner. I say :+1: after fixing the docs.

isratrade commented:
@eLobato, pushed separate commit per your comments. We are at 128 comments for this PR!

orrabin commented:
@isratrade I went over your last commit and there are only a few nitpicks left. There are a few places where it still says `ptables` instead on `partition table`.

isratrade commented:
@orrabin, fixed per your comments. @eLobato, squashed and rebased. It should be ready for merge if all tests pass.

dLobatog commented on app/controllers/api/v2/architectures_controller.rb:
```diff
@@ -1,16 +1,16 @@
       api :GET, "/architectures/", N_("List all architectures")
+      api :GET, "/operatingsystems/:operatingsystem_id/architectures", N_("List all architectures for operating system")
       param_group :search_and_pagination, ::Api::V2::BaseController
+      param :operatingsystem_id, String, :desc => N_("ID of operatingsystem")
```
`operating system`

dLobatog commented on app/controllers/api/v2/environments_controller.rb:
```diff
@@ -5,16 +5,20 @@ class EnvironmentsController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy}
 
       api :GET, "/environments/", N_("List all environments")
+      api :GET, "/puppetclasses/:puppetclass_id/environments", N_("List environments of puppet class")
```
`Puppet`, capitalized for consistency with the rest of the docs.

dLobatog commented on app/controllers/api/v2/hostgroups_controller.rb:
```diff
@@ -5,17 +5,19 @@ class HostgroupsController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy clone}
 
       api :GET, "/hostgroups/", N_("List all host groups")
+      api :GET, "/puppetclasses/:puppetclass_id/hostgroups", N_("List all host groups for a puppet class")
```
Puppet

dLobatog commented on app/controllers/api/v2/hostgroups_controller.rb:
```diff
@@ -5,17 +5,19 @@ class HostgroupsController < V2::BaseController
+      api :GET, "/puppetclasses/:puppetclass_id/hostgroups", N_("List all host groups for a puppet class")
+      api :GET, "/locations/:location_id/hostgroups", N_("List all host groups per location")
+      api :GET, "/organizations/:organization_id/hostgroups", N_("List all host groups per organization")
+      param :puppetclass_id, String, :desc => N_("ID of puppet class")
```
Puppet

dLobatog commented:
@isratrade tests pass, thanks. There are 4 or 5 docs that don't look as the rest of the API docs then I'll merge it, it's a great add. Sorry if you got many annoying emails with one word, I commented every typo I found to make it easier to fix.

Commit:
fixes #3492 - API v2 nested routes for each controller

isratrade commented:
@eLobato, fixed per your comments, plus more instances of hostgroup to host group, puppet class to Puppet class, and operatingsystem to operating system

dLobatog commented:
[test] , (after the recent Rubocop fixes)

dLobatog commented:
Thanks, this is a major cleanup! :smile:

Ready to merge if Jenkins is green. I'll follow up with a PR to auto-generate `resource_scope_for_index` since it seems it only needs to be overridden in a few places.

dLobatog commented on app/controllers/api/v2/os_default_templates_controller.rb:
```diff
@@ -9,12 +9,13 @@ class OsDefaultTemplatesController < V2::BaseController
       before_filter :find_resource, :only => %w{show update destroy}
 
       api :GET, '/operatingsystems/:operatingsystem_id/os_default_templates', N_('List default templates combinations for an operating system')
+      api :GET, '/config_templates/:config_template_id/os_default_templates', N_('List os default templates for provisioning template')
```
I'll change this `os default templates`  to `List operating systems where this template is set as a default` on merge.

dLobatog commented on app/controllers/api/v2/subnets_controller.rb:
```diff
@@ -5,17 +5,19 @@ class SubnetsController < V2::BaseController
 
       api :GET, '/subnets', N_("List of subnets")
+      api :GET, "/domains/:domain_id/subnets", N_("List of subnets for a domain")
+      api :GET, "/locations/:location_id/subnets", N_("List of domains per location")
```
`List of subnets per location`.

dLobatog commented on app/controllers/api/v2/subnets_controller.rb:
```diff
@@ -5,17 +5,19 @@ class SubnetsController < V2::BaseController
       api :GET, '/subnets', N_("List of subnets")
+      api :GET, "/domains/:domain_id/subnets", N_("List of subnets for a domain")
+      api :GET, "/locations/:location_id/subnets", N_("List of domains per location")
+      api :GET, "/organizations/:organization_id/subnets", N_("List of domains per organization")
```
`List of subnets per organization`.

dLobatog commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -8,14 +8,23 @@ class UsersController < V2::BaseController
+      before_filter :find_resource, :only => [:show, :update, :destroy]
 
       api :GET, "/users/", N_("List all users")
+      api :GET, "/auth_source_ldaps/:auth_source_ldap_id/users", N_("List all users for auth source LDAP")
```
`LDAP authentication source`

dLobatog commented on app/controllers/api/v2/users_controller.rb:
```diff
@@ -8,14 +8,23 @@ class UsersController < V2::BaseController
+      api :GET, "/roles/:role_id/users", N_("List all users for role")
+      api :GET, "/locations/:location_id/users", N_("List all users for location")
+      api :GET, "/organizations/:organization_id/users", N_("List all users for organization")
+      param :auth_source_ldap_id, String, :desc => N_("ID of auth source LDAP")
```
`ID of LDAP authentication source`

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,20 +5,23 @@ class ConfigTemplatesController < V2::BaseController
       before_filter :process_operatingsystems, :only => [:create, :update]
 
       api :GET, "/config_templates/", N_("List provisioning templates")
+      api :GET, "/operatingsystem/:operatingsystem_id/config_templates", N_("List templates per operating system")
```
`List provisioning templates`

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,20 +5,23 @@ class ConfigTemplatesController < V2::BaseController
 
       api :GET, "/config_templates/", N_("List provisioning templates")
+      api :GET, "/operatingsystem/:operatingsystem_id/config_templates", N_("List templates per operating system")
+      api :GET, "/locations/:location_id/config_templates/", N_("List templates per location")
```
`List provisioning templates`

dLobatog commented on app/controllers/api/v2/config_templates_controller.rb:
```diff
@@ -5,20 +5,23 @@ class ConfigTemplatesController < V2::BaseController
       api :GET, "/config_templates/", N_("List provisioning templates")
+      api :GET, "/operatingsystem/:operatingsystem_id/config_templates", N_("List templates per operating system")
+      api :GET, "/locations/:location_id/config_templates/", N_("List templates per location")
+      api :GET, "/organizations/:organization_id/config_templates/", N_("List templates per organization")
```
`List provisioning templates`

dLobatog commented:
@isratrade I've sent you a PR to your branch https://github.com/isratrade/foreman/issues/4 with the (hopefully) last fixes, please merge and check if tests are still green and I'll merge.

isratrade commented:
@eLobato, merged you patch with a minor change. thanks!

dLobatog commented on app/controllers/api/v2/external_usergroups_controller.rb:
```diff
@@ -7,14 +7,15 @@ class ExternalUsergroupsController < V2::BaseController
       before_filter :find_required_nested_object, :only => [:index, :show, :create]
 
       api :GET, '/usergroups/:usergroup_id/external_usergroups', N_('List all external user groups for user group')
+      api :GET, '/auth_source_ldaps/:auth_source_ldap_id/external_usergroups', N_('List all external user groups for auth source')
```
LDAP authentication source

dLobatog commented on app/controllers/api/v2/external_usergroups_controller.rb:
```diff
@@ -7,14 +7,15 @@ class ExternalUsergroupsController < V2::BaseController
       end
 
       api :GET, '/usergroups/:usergroup_id/external_usergroups/:id', N_('Show an external user group for user group')
+      api :GET, '/auth_source_ldaps/:auth_source_ldap_id/external_usergroups/:id', N_('Show an external user group for auth source')
```
LDAP authentication source

dLobatog commented:
Merged as f4f1ff0042943b74b61468d03e32771a34c9722d and 453dc6939f607d8ae0ff4580b5562734c60b19c5 , thanks @isratrade !
